Create webhooks automatically

qaisjp opened the pull request:
## Summary

Fixes #764. Related to #1255, #1296, #1322.

I'm submitting it now to get a code review. I will do more detailed testing after any requested code changes.

- **Removed** `WebhookURL` global setting is removed and will quit matterbridge
- **New** `AutoWebhooks=true` setting, which will automatically use (and create, if they do not exist) webhooks inside specific channels. This only works if the bot has `Manage Webhooks` permission in bridged channels (global permission or as a channel permission override)
- **No change** Backwards compatibility with channel-specific webhooks

## Test plan

**With** `AutoWebhooks=true`:

- with global permission:
    1. A single channel has existing webhook picked up
    2. A single channel that has a defined WebhookURL is used, and other channels pick up existing webhooks
    3. Same (2) except, on message send, it creates for channels that don't have a webhook
- without global permission
    1. Works in a channel with a defined WebhookURL
    2. Picks up in channels that have Manage Webhook permission
    3. Bridge shuts down with an error message if it does not have permission to access that channel

**With** `AutoWebhooks=false`:

- with and without global permission
    1. A single channel, with a defined WebhookURL, works
    2. (1) works, and other channels (without WebhookURL) use regular bot messages

42wim commented on bridge/discord/transmitter/transmitter.go:
```diff
@@ -0,0 +1,247 @@
+	wh := t.getWebhook(channelID)
+
+	// If we don't have a webhook for this channel...
+	if wh == nil {
+		// Early exit, if we don't want to automatically create one
+		if !t.autoCreate {
+			return nil, ErrWebhookNotFound
+		}
+
+		// Try and create one
+		t.Log.Infof("Creating a webhook for %s\n", channelID)
+		var err error
+		wh, err = t.createWebhook(channelID)
+		if err != nil {
+			return nil, fmt.Errorf("could not create webhook: %w", err)
+		}
+	}
```
maybe this code can be put in to the `getWebhook` which can return a webhook or an error

qaisjp replied:
I moved this into a `getOrCreateWebhook` function instead, as `getWebhook` is also used by `Edit` and just handles the mutex and gets it out of the map.

Done in 70aaf3a

42wim commented on bridge/discord/transmitter/transmitter.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// AddWebhook allows you to register a channel's webhook with the transmitter.
+func (t *Transmitter) AddWebhook(channelID string, webhook *discordgo.Webhook) (replaced bool) {
```
```suggestion
func (t *Transmitter) AddWebhook(channelID string, webhook *discordgo.Webhook) bool {
```

42wim commented on bridge/discord/transmitter/transmitter.go:
```diff
@@ -0,0 +1,247 @@
+	}
+
+	// Get all existing webhooks
+	hooks, err := t.session.GuildWebhooks(t.guild)
+	manuallyFetchHooks := len(channelIDs) > 0
```
I'd put this line above the `hooks, err := t.session.GuildWebhooks(t.guild)` as it's more idiomatic to immediately handle an error

42wim replied:
```suggestion
	manuallyFetchHooks := len(channelIDs) > 0
	hooks, err := t.session.GuildWebhooks(t.guild)
```

42wim commented on bridge/discord/transmitter/transmitter.go:
```diff
@@ -0,0 +1,247 @@
+	hooks, err := t.session.GuildWebhooks(t.guild)
+	manuallyFetchHooks := len(channelIDs) > 0
+
+	// Check to make sure we have permissions
+	if isDiscordPermissionError(err) {
+		// If we shouldn't try manually fetching hooks, return an error
+		if !manuallyFetchHooks {
+			return fmt.Errorf("the 'Manage Webhooks' permission is required: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("could not get webhooks: %w", err)
+	}
```
```suggestion
if err != nil {
        // Check to make sure we have permissions
        // If we shouldn't try manually fetching hooks, return an error
	if isDiscordPermissionError(err) && !manuallyFetchHooks {
		return fmt.Errorf("the 'Manage Webhooks' permission is required: %w", err)
	}

	return fmt.Errorf("could not get webhooks: %w", err)
}
```

42wim replied:
this makes it more readable

qaisjp replied:
I'll try something like this, this won't immediately work, as we don't want to return an error when `isDiscordPermissionError(err)` and `manuallyFetchHooks` is true.

(This is a global permission error, and tries per-channel permissions as a fallback)

42wim commented on bridge/discord/webhook.go:
```diff
@@ -0,0 +1,99 @@
+// account or protocol, and if so, returns the Discord avatar (if exists)
+func (b *Bdiscord) maybeGetLocalAvatar(msg *config.Message) string {
+	for _, val := range b.GetStringSlice("UseLocalAvatar") {
+		if msg.Protocol == val || msg.Account == val {
+			member, err := b.getGuildMemberByNick(msg.Username)
+			if err != nil {
+				return ""
+			}
+			return member.User.AvatarURL("")
```
```suggestion
		if msg.Protocol != val && msg.Account != val {
		        continue
		}
			
		member, err := b.getGuildMemberByNick(msg.Username)
		if err != nil {
			return ""
		}
		
		return member.User.AvatarURL("")
```

42wim reviewed (commented):
Haven't tested it yet, some general styling comments

qaisjp commented on bridge/discord/discord.go:
```diff
@@ -227,23 +242,9 @@ func (b *Bdiscord) Send(msg config.Message) (string, error) {
 	// Use webhook to send the message
-	if wID != "" && msg.Event != config.EventMsgDelete {
+	useWebhooks := b.shouldMessageUseWebhooks(&msg)
+	if useWebhooks && msg.Event != config.EventMsgDelete {
```
Done in b89f0fd

qaisjp commented:
let me know if you're happy with 5595e4a513c6074553274c04185566623a9f4d2b and I'll merge that into a243f79

qaisjp commented:
I'm also running this in prod now so hopefully that will catch any problems

qaisjp commented:
- Let me know if you're happy with 5595e4a513c6074553274c04185566623a9f4d2b and I'll merge that into a243f79
- I'm also running this in prod now so hopefully that will catch any problems
- I've redone the tests in the OP (and I've completed manual testing)

42wim commented on bridge/discord/discord.go:
```diff
@@ -137,40 +140,52 @@ func (b *Bdiscord) Connect() error {
+		message := "The global WebhookURL setting has been removed. "
+		message += "You can get similar \"webhook editing\" behaviour by replacing this line with `AutoWebhooks=true`. "
+		message += "If you rely on the old-OLD (non-editing) behaviour, can move the WebhookURL to specific channel sections."
+		b.Log.Fatalln(message)
```
Could you print this out as a b.Log.Errorln and return an error instead of doing fatal there?
Matterbridge self will do the log.Fatal when it bubbles up. WDYT?

42wim commented on bridge/discord/helpers.go:
```diff
@@ -204,7 +204,7 @@ func (b *Bdiscord) splitURL(url string) (string, string) {
 	webhookURLSplit := strings.Split(url, "/")
 	if len(webhookURLSplit) != expectedWebhookSplitCount {
-		b.Log.Fatalf("%s is no correct discord WebhookURL", url)
+		b.Log.Fatalf("%s is an incorrect discord WebhookURL", url)
```
I know it was already here, but it would be cleaner if we can have this bubble up instead of fatal here.
WDYT?

42wim commented:
> * Let me know if you're happy with [5595e4a](https://github.com/42wim/matterbridge/commit/5595e4a513c6074553274c04185566623a9f4d2b) and I'll merge that into [a243f79](https://github.com/42wim/matterbridge/commit/a243f795c3a959e9eed11fa5f6bcc0c388ef5c05)

I can live with it, it's all very subjective :)
(I'm more a fan of https://github.com/42wim/matterbridge/pull/1323#discussion_r537046553 as I find it more readable and personally don't like the `err == nil` statement)

Haven't tested it yet, but LGTM. (will try tomorrow)
2 small nitpicks about errors.

Thanks! 👍

qaisjp commented:
I think I've addressed your nits, let me know if you're happy with it!

I've also tried a couple different formats for that error handling, please can you help find the best one?

(Alt formatting 3 didn't pass golangci-lint as the `if err != nil` contents were too deeply nested.)

42wim commented on bridge/discord/transmitter/transmitter.go:
```diff
@@ -0,0 +1,252 @@
+	}
+
+	// Get all existing webhooks
+	hooks, err := t.session.GuildWebhooks(t.guild)
+	if err != nil {
+		switch {
+		case isDiscordPermissionError(err):
+			// We fallback on manually fetching hooks from individual channels
+			// if we don't have the "Manage Webhooks" permission globally.
+			// We can only do this if we were provided channelIDs, though.
+			if len(channelIDs) == 0 {
+				return ErrPermissionDenied
+			}
+			t.Log.Debugln("Missing global 'Manage Webhooks' permission, falling back on per-channel permission")
+		default:
+			return fmt.Errorf("could not get webhooks: %w", err)
+		}
+	} else {
+		t.Log.Debugln("Refreshing guild webhooks using global permission")
+		t.assignHooksByAppID(hooks, botID, false)
+		return nil
+	}
```
What about creating a separate helper function `t.getChannelHooks(channelIDs []string) error` for all the code coming after this if clause ?

Then this can become clearer without the need for an else.

```suggestion
	hooks, err := t.session.GuildWebhooks(t.guild)
	if err != nil {
		switch {
		case isDiscordPermissionError(err):
			// We fallback on manually fetching hooks from individual channels
			// if we don't have the "Manage Webhooks" permission globally.
			// We can only do this if we were provided channelIDs, though.
			if len(channelIDs) == 0 {
				return ErrPermissionDenied
			}
			t.Log.Debugln("Missing global 'Manage Webhooks' permission, falling back on per-channel permission")
			return t.getChannelHooks(channelIDs []string)
		default:
			return fmt.Errorf("could not get webhooks: %w", err)
		}
	} 	
	
	t.Log.Debugln("Refreshing guild webhooks using global permission")
	t.assignHooksByAppID(hooks, botID, false)
	
	return nil
```

qaisjp replied:
Hmm I think that clashes with `getWebhook` which currently takes a channel ID and returns the corresponding `*discordgo.Webhook`. Is `fetchChannelsHooks` okay? I've gone with that in the latest push

42wim replied:
sure, the name was just a suggestion :)

42wim commented:
🎉
